### PR TITLE
SPARK-6541 - Sort executors by ID (numeric)

### PR DESCRIPTION
**Pull request by jbonofre**
"Force" the executor ID sort with Int.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala [hunk 1]
```diff
@@ -43,8 +43,9 @@ private[ui] class ApplicationPage(parent: MasterWebUI) extends WebUIPage("app")
       return UIUtils.basicSparkPage(msg, "Not Found")
     }
 
-    val executorHeaders = Seq("ExecutorID", "Worker", "Cores", "Memory", "State", "Logs")
+    val executorHeaders = Seq("Executor ID", "Worker", "Cores", "Memory", "State", "Logs")
     val allExecutors = (app.executors.values ++ app.removedExecutors).toSet.toSeq
+      .sortWith(_.id < _.id)
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala]: Can this be `sortBy(_.id)`? and I assume it's already ints.

File: core/src/main/scala/org/apache/spark/ui/jobs/ExecutorTable.scala [hunk 2]
```diff
@@ -102,7 +102,7 @@ private[ui] class ExecutorTable(stageId: Int, stageAttemptId: Int, parent: Stage
 
     listener.stageIdToData.get((stageId, stageAttemptId)) match {
       case Some(stageData: StageUIData) =>
-        stageData.executorSummary.toSeq.sortBy(_._1).map { case (k, v) =>
+        stageData.executorSummary.toSeq.sortWith(_._1.toInt < _._1.toInt).map { case (k, v) =>
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/ui/jobs/ExecutorTable.scala]: Can also be `sortBy(_._1.toInt)` I think. Also, are we sure these fit into 32-bit ints?

Comment (jbonofre) [inline core/src/main/scala/org/apache/spark/ui/jobs/ExecutorTable.scala]: Good point, toLong makes sense. Let me update.

File: core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala (hunk 1 above)
Comment (jbonofre) [inline core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala]: Yes, here, it's already an Int. Let me update for sortBy() instead of sortWith().

Comment (jbonofre) [inline core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala]: I'm not sure we are able to use sortBy() there as it can diverge to ordering. Let me double check.

Comment (jbonofre) [inline core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala]: I think I have to cast toIndexedSeq for the ordering with sortBy (instead of toSeq). Let me try.

Comment (jbonofre) [inline core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala]: I can't use sortBy() due to toSeq (and so diverging implicit expansion for ordering). I keep sortWith() for now (or I have to change the toSeq statement).

Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala]: Don't you just need a `toList` then? indeed a `Seq` may not be random-access and so hard to sort.

Comment (jbonofre) [inline core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala]: @srowen yes, it's what I thought. Let me update like this. Thanks (as usual ;))

Comment (squito) [inline core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala]: @srowen actually Scala's `List` is a linked list.  You may want `toIndexedSeq`.  In any case, I think all of them will dump the data into an array for sorting (even `ArrayBuffer`), so it probably doesn't matter.

Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala]: Fair enough, though I wonder then why the `Seq` didn't work. Not a big deal but I can't see a good semantic reason for `sortBy` to not work but `sortWith` to work.

Comment (squito) [inline core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala]: yeah, that doesn't make sense to me either, I don't see any reason why `toSeq` wouldn't cut it -- seems to compile for me ...

**Comment by andrewor14** [comment]
ok to test

**Comment by srowen** [comment]
I'm OK with this. Sorting executor IDs numerically seems like the intent anyway. I don't see a downside to sorting the display otherwise, other than the cost of sorting, but that's pretty trivial given the number of executors.

File: core/src/main/scala/org/apache/spark/ui/jobs/ExecutorTable.scala [hunk 3]
```diff
@@ -102,7 +102,7 @@ private[ui] class ExecutorTable(stageId: Int, stageAttemptId: Int, parent: Stage
 
     listener.stageIdToData.get((stageId, stageAttemptId)) match {
       case Some(stageData: StageUIData) =>
-        stageData.executorSummary.toSeq.sortBy(_._1).map { case (k, v) =>
+        stageData.executorSummary.toSeq.sortBy(_._1.toLong).map { case (k, v) =>
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/ui/jobs/ExecutorTable.scala]: Are executor ids always guaranteed to be longs? If not, this could fail with a NumberFormatException.

Comment (srowen) [inline core/src/main/scala/org/apache/spark/ui/jobs/ExecutorTable.scala]: Oh, it's just occurred to me that the driver is an executor with ID "driver" for example. @jbonofre did that not actually cause a problem?

Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/ui/jobs/ExecutorTable.scala]: Why not just use JS to trigger the sort in the frontend or to set a default sort order for the table? Are you worried about UI flickering?

Comment (jbonofre) [inline core/src/main/scala/org/apache/spark/ui/jobs/ExecutorTable.scala]: @JoshRosen yes, it's an alternative. Let me update like this.

Comment (srowen) [inline core/src/main/scala/org/apache/spark/ui/jobs/ExecutorTable.scala]: @jbonofre are you still working on this one?

Comment (jbonofre) [inline core/src/main/scala/org/apache/spark/ui/jobs/ExecutorTable.scala]: I'm in vacation up to Wednesday. So I will resume my work on this on Wednesday night. Sorry for the delay.

Comment (jbonofre) [inline core/src/main/scala/org/apache/spark/ui/jobs/ExecutorTable.scala]: I'm back on this PR. I'm resume my work on it using JS for sorting (instead of server side).

Comment (jbonofre) [inline core/src/main/scala/org/apache/spark/ui/jobs/ExecutorTable.scala]: I just checked the usage of sorttable (from kryogenix) used in Spark UI, and actually, we have to sort server side.
Basically, using sorttable js at page loading is wrong. Here's the extract for the sorttable FAQ:

"
Sorting the table when the page is loaded

Lots of people ask, "how do I make sorttable sort the table the first time the page is loaded?" The answer is: you don't. Sorttable is about changing the HTML that is served from your server without a page refresh. When the page is first served from the server, you have to incur the wait for it to be served anyway. So, if you want the table sorted when a page is first displayed, serve the table in sorted order. Tables often come out of a database; get the data from the database in a sorted order with an ORDER BY clause in your SQL. Any solution which involves you running sorttable as soon as the page loads (i.e., without user input) is a wrong solution.
"

So, according to @JoshRosen (good catch Josh ;)), executor ID is not necessary a long (for instance, "driver" executor). Anyway, let me check if I can sort (alphanumeric) on server side, matching the sorttable sort ordering.

Comment (srowen) [inline core/src/main/scala/org/apache/spark/ui/jobs/ExecutorTable.scala]: Hm, that's not a great argument to me, as it means "you must implement sorting logic twice, always". It's not possible to trigger the JS after the table loads? in any event, it may not be so hard to sort both places, sure. But it's a weird sort: by number for things that can be numbers but then alphabetically otherwise. That has to be implemented two places. It may prove to be too much hassle, as in fact, executors aren't really numbers but only usually appear to be

Comment (jbonofre) [inline core/src/main/scala/org/apache/spark/ui/jobs/ExecutorTable.scala]: I agree @srowen. I can call the sorttable sorting with something like onload. Let me try this.

Comment (jbonofre) [inline core/src/main/scala/org/apache/spark/ui/jobs/ExecutorTable.scala]: It seems that the innerSortFunction function requires latest sorttable version. Let me try to upgrade.

Comment (jbonofre) [inline core/src/main/scala/org/apache/spark/ui/jobs/ExecutorTable.scala]: I'm testing the usage of:

    <script>
      {
        Unparsed {
          """
            |      window.onload = function() {
            |        sorttable.innerSortFunction.apply(document.getElementById('executorid'), [])
            |      };
          """.stripMargin
        }
      }
    </script>

in the ExecutorTable, but also in the UIUtils generate table function (as used in stage view) to see if it works as expected.

**Comment by jbonofre** [comment]
Update with client side executor ID sorting. This PR answers the original request, but I think it makes sense to sort all executor tables the same way.

**Comment by jbonofre** [comment]
Thoughts ?

File: core/src/main/resources/org/apache/spark/ui/static/sorttable.js [hunk 4]
```diff
@@ -3,19 +3,19 @@
   version 2
   7th April 2007
   Stuart Langridge, http://www.kryogenix.org/code/browser/sorttable/
-  
+
```
Comment (srowen) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js]: It's not a big deal, but stripping whitespace on lines just adds to the PR noise and possibility of merge conflicts. If you can disable this it's much preferred.

File: core/src/main/resources/org/apache/spark/ui/static/sorttable.js [hunk 5]
```diff
@@ -81,18 +81,18 @@ sorttable = {
...
-        if (mtch && typeof sorttable["sort_"+override] == 'function') {
-          headrow[i].sorttable_sortfunction = sorttable["sort_"+override];
-        } else {
-          headrow[i].sorttable_sortfunction = sorttable.guessType(table,i);
-        }
-        // make it clickable to sort
-        headrow[i].sorttable_columnindex = i;
-        headrow[i].sorttable_tbody = table.tBodies[0];
-        dean_addEvent(headrow[i],"click", function(e) {
+	      if (mtch && typeof sorttable["sort_"+override] == 'function') {
```
Comment (srowen) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js]: This adds tab-based indentation instead of spaces; later extra space indents are used where I think the indentation was correct. You might review and/or back out whatever superfluous changes in whitespace that you can.

File: core/src/main/resources/org/apache/spark/ui/static/sorttable.js [hunk 6]
```diff
@@ -100,20 +100,20 @@ sorttable = {
             this.removeChild(document.getElementById('sorttable_sortfwdind'));
             sortrevind = document.createElement('span');
             sortrevind.id = "sorttable_sortrevind";
-            sortrevind.innerHTML = stIsIE ? '&nbsp<font face="webdings">5</font>' : '&nbsp;&#x25BE;';
+            sortrevind.innerHTML = stIsIE ? '&nbsp<font face="webdings">5</font>' : '&nbsp;&#x25B4;';
```
Comment (srowen) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js]: So this is now an up arrow not down arrow -- just making sure I read the intend of the change correctly

**Comment by srowen** [comment]
Hm, by the end here it seems like 90% of the changes are whitespace changes. It's enough compared to the small functional change that I think it's worth re-doing.

In the end how does this sort? numerically, but what about IDs like "driver"? does it sort on something like the leading prefix of digits?

Does it still need server-side sorting if so? I see that's added but it just sorts lexicographically.

**Comment by jbonofre** [comment]
Actually, I didn't change the file myself: I just upgraded to latest sorttable.js script (downloaded from the website). That explains the whitespace changes. I can try to clean it up.

The ordering is only client-side (no change on the server side), and it use numeric/alphanumeric sort. As you can see in the Jira, the user would like ordering (as when he clicks on table header) instead of the "default" one.

Thanks Sean !

**Comment by srowen** [comment]
Oh OK, then let's keep the whitespace changes, if they're part of the upstream code we're including. I wouldn't change it then. 

There is still a sorting change in `ApplicationPage.scala` right?
What's the nature of the sort -- how do the following sort, just theoretically and out of curiosity:
- driver
- 1
- 10
- driver1
- driver10
- 1driver
- 10driver

**Comment by jbonofre** [comment]
Let me prepare couple of screenshots to illustrate the sort order.

**Comment by jbonofre** [comment]
PR rebased/cleaned up/updated.

**Comment by jbonofre** [comment]
Here's the screenshot without the PR:
screen2

and here's the screenshot with the PR:
screen

Note the default sorting.

**Comment by srowen** [comment]
Yeah I'm still curious where "driver" would sort for example? I think it's clear that sorting numbers as numbers is OK but I don't see how this handles non-numeric IDs. Any reasonable behavior seems fine; just want to make sure it works.

**Comment by jbonofre** [comment]
Yeah, I understand what you mean. Let me simulate the case and attach a screenshot. Thanks again Sean !

**Comment by jbonofre** [comment]
PR rebased. Test failure seems related to flaky tests (Hive). Anyway, I'm trying on my box.

File: core/src/main/resources/org/apache/spark/ui/static/sorttable.js (hunk 5 above)
Comment (andrewor14) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js]: yeah can you use spaces instead?

File: core/src/main/resources/org/apache/spark/ui/static/sorttable.js [hunk 7]
```diff
@@ -134,35 +134,35 @@ sorttable = {
...
-          }
-        
-          delete row_array;
-        });
-      }
+	        // build an array to sort. This is a Schwartzian transform thing,
+	        // i.e., we "decorate" each row with the actual sort key,
+	        // sort based on the sort keys, and then put the rows back in order
+	        // which is a lot faster because you only do getInnerText once per row
+	        row_array = [];
```
Comment (andrewor14) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js]: these are indented too much. Also please uses spaces where possible

File: core/src/main/scala/org/apache/spark/ui/jobs/ExecutorTable.scala [hunk 8]
```diff
@@ -89,6 +89,15 @@ private[ui] class ExecutorTable(stageId: Int, stageAttemptId: Int, parent: Stage
...
     </table>
+      <script>
+        {Unparsed {
+        """
+          |      window.onload = function() {
+          |        sorttable.innerSortFunction.apply(document.getElementById('executorid'), [])
+          |      };
+        """.stripMargin
+      }}
+      </script>
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ui/jobs/ExecutorTable.scala]: style:
```
</table>
<script>
  {Unparsed {
    ...
  }}
</script>
```

**Comment by andrewor14** [comment]
@jbonofre actually I do mind the whitespace changes. Elsewhere we try to use 2 space indentations in JavaScript instead of tabs. This PR currently makes the style inconsistent everywhere. Would you mind fixing these manually?

**Comment by jbonofre** [comment]
Sure, let me clean it up.

**Comment by jbonofre** [comment]
PR rebased and cleaned up (about indentation and style)

**Comment by jbonofre** [comment]
Rebase

File: core/src/main/resources/org/apache/spark/ui/static/sorttable.js [hunk 9]
```diff
@@ -169,7 +169,7 @@ sorttable = {
     for (var i=0; i<table.tBodies[0].rows.length; i++) {
       text = sorttable.getInnerText(table.tBodies[0].rows[i].cells[column]);
       if (text != '') {
-        if (text.match(/^-?[�$�]?[\d,.]+%?$/)) {
+        if (text.match(/^-?[�$�]?[\d,.]+%?$/)) {
```
Comment (srowen) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js]: The special characters seem to have changed here -- is that intended or an accidental change?

**Comment by srowen** [comment]
I think I'm still curious about what ordering this defines and whether it works with non-numeric IDs? I think this is close to done otherwise.

**Comment by jbonofre** [comment]
Hi Sean. Yes, let me prepare some "dummy" data to test different ordering. Thanks !

**Comment by jbonofre** [comment]
Rebase.

**Comment by jbonofre** [comment]
Some quick screenshots to show the ordering depending of the executor ID (without the sorting, we have the table rows in any order).

Sorting asc:
screen1

Sorting desc:
screen2

**Comment by jbonofre** [comment]
Rebased.

**Comment by jbonofre** [comment]
PR rebased.

**Comment by srowen** [comment]
LGTM

**Comment by jbonofre** [comment]
Thanks Sean ! Let me know if you need anything.

**Comment by jbonofre** [comment]
PR rebased

File: core/src/main/resources/org/apache/spark/ui/static/sorttable.js (hunk 9 above)
Comment (srowen) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js]: @jbonofre wanted to triple-check on this -- is this change intended?

Comment (jbonofre) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js]: My bad Sean, you are right: it's not intended (resulting upgrade to latest sorttable version). Let me fix that.

**Comment by jbonofre** [comment]
Rebase

File: core/src/main/resources/org/apache/spark/ui/static/sorttable.js [hunk 10]
```diff
@@ -169,7 +169,7 @@ sorttable = {
     for (var i=0; i<table.tBodies[0].rows.length; i++) {
       text = sorttable.getInnerText(table.tBodies[0].rows[i].cells[column]);
       if (text != '') {
-        if (text.match(/^-?[�$�]?[\d,.]+%?$/)) {
+        if (text.match(/^-?[£$¤]?[\d,.]+%?$/)) {
```
Comment (srowen) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js]: Heh, github says this line changed but I don't see it. I'm worried it's a difference in some unprintable character that will throw this off. Any way to make sure this line really didn't change? copy and paste from the previous version or is that what you did?

Comment (jbonofre) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js]: Let me checkout from master and change just the innerStoreFunction.

**Comment by jbonofre** [comment]
Cleanup and rebase

**Comment by srowen** [comment]
All looks good now. And it still works ? :)

**Comment by jbonofre** [comment]
Lol ;) Let me prepare a screenshot ;)

**Comment by jbonofre** [comment]
Last screenshot showing that it still works ;)

screen

**Comment by jbonofre** [comment]
Rebase

**Comment by srowen** [comment]
It's looking good. If there are no further comments I'll merge shortly.

**Comment by jbonofre** [comment]
Awesome, thanks Sean !
